src/python/pants/backend/python/rules/run_setup_py.py

```diff
@@ -128,12 +136,94 @@ class SetupPySources:
   package_data: Tuple[PackageDatum, ...]
 
 
+@dataclass(frozen=True)
+class SetupPyChrootRequest:
+  """A request to create a chroot containing a setup.py and the sources it operates on."""
+  exported_target: ExportedTarget
+  py2: bool = False  # Whether to use py2 or py3 namespace package semantics.
+
+
+@dataclass(frozen=True)
+class SetupPyChroot:
+  """A chroot containing a generated setup.py and the sources it operates on."""
+  digest: Digest
+  # The keywords are embedded in the setup.py file in the digest, so these aren't
+  # strictly needed here, but they are convenient for testing.
+  setup_keywords_json: str
+
+
 @dataclass(frozen=True)
 class SetuptoolsSetup:
   """The setuptools tool."""
   requirements_pex: Pex
 
 
+# We write .py sources into the chroot under this dir.
+CHROOT_SOURCE_ROOT = 'src'
+
+
+SETUP_BOILERPLATE = """
+# DO NOT EDIT THIS FILE -- AUTOGENERATED BY PANTS
+# Target: {target_address_spec}
+
+from setuptools import setup
+
+setup(**{setup_kwargs_str})
+"""
+
+@rule
+async def generate_chroot(request: SetupPyChrootRequest) -> SetupPyChroot:
+  if request.py2:
+    # TODO: Implement Python 2 support.  This will involve, among other things: merging ancestor
+    # __init__.py files into the chroot, detecting packages based on the presence of __init__.py,
+    # and inspecting all __init__.py files for the namespace package incantation.
+    raise UnsupportedPythonVersion('Running setup.py commands not supported for Python 2.')
+
+  owned_deps = await Get[OwnedDependencies](DependencyOwner(request.exported_target))
+  targets = HydratedTargets(od.hydrated_target for od in owned_deps)
+  sources = await Get[SetupPySources](SetupPySourcesRequest(targets))
+  requirements = await Get[ExportedTargetRequirements](DependencyOwner(request.exported_target))
+
+  # Nest the sources under the src/ prefix.
+  src_digest = await Get[Digest](
+    DirectoryWithPrefixToAdd, DirectoryWithPrefixToAdd(sources.digest, CHROOT_SOURCE_ROOT))
+
+  # Generate the kwargs to the setup() call.
+  setup_kwargs = request.exported_target.hydrated_target.adaptor.provides.setup_py_keywords.copy()
+  setup_kwargs.update({
+    'package_dir': {'': CHROOT_SOURCE_ROOT},
+    'packages': sources.packages,
+    'namespace_packages': sources.namespace_packages,
+    'package_data': dict(sources.package_data),
+    'install_requires': sorted(requirements.requirement_strs)
```

Comment:
This sorting does not respect declration order which is sometimes significant and is in contrast to v1. The surprise this creates would be getting a different resolve for `./pants repl <exported dist target>` for example from `pip install <exported dist>`. The `requirements.requirement_strs` should be stable so I think the sorting can just be eliminated.

Reply:
Good catch. In fact, I was also sorting on creation of the ExportedTargetRequirements instance, which is also wrong. Removed that as well. Now the order will be whatever ` PexRequirements.create_from_adaptors()` returns, with the sorted exported target reqs appended. I assume `PexRequirements.create_from_adaptors()` order is stable?